### PR TITLE
Fix corner case in router fuson

### PR DESCRIPTION
**Pull request by Autumn1998**
# Description

Fix kernels bug on the case: all the logits < 0

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/fused_router/utils.h [hunk 1]
```diff
@@ -57,13 +71,22 @@ __device__ inline void apply_sigmoid_on_float(DataType *scores, int data_size, i
...
+                                                ReduceFuncType type, int lane_id) {
+  T (*reduce_func)(T, T);
+  double default_val = 0;
+  if (type == ReduceFuncType::SUM) {
+    reduce_func = sum;
+    default_val = 0;
+  } else if (type == ReduceFuncType::MAX) {
+    reduce_func = max;
+    default_val = -1e30;
+  }
```
Comment (yaox12) [inline transformer_engine/common/fused_router/utils.h:83]: Use `-std::numeric_limits<double>::infinity()`?

File: transformer_engine/common/fused_router/utils.h [hunk 2]
```diff
@@ -26,14 +26,28 @@ __device__ inline T sum(T a, T b) {
...
+__device__ inline T warp_reduce_on_shmem(T *data_ptr, int data_size, ReduceFuncType type,
                                          int lane_id) {
+  T (*reduce_func)(T, T);
+  double default_val = 0;
+  if (type == ReduceFuncType::SUM) {
+    reduce_func = sum;
+    default_val = 0;
+  } else if (type == ReduceFuncType::MAX) {
+    reduce_func = max;
+    default_val = -1e30;
```
Comment (yaox12) [inline transformer_engine/common/fused_router/utils.h]: Use `-std::numeric_limits<double>::infinity()`?

File: transformer_engine/common/fused_router/utils.h [hunk 3]
```diff
@@ -145,19 +170,29 @@ __device__ inline void apply_softmax_on_float(DataType *scores, int data_size, i
...
+  };
   // Topk Times: Find the max value and its index
   // Then mask it, and record the index in the topk_indices
   // After looping topk times, the topk_indices will be the topk indices
   for (int k = 0; k < topk; k++) {
     // Find the max value and its index
-    volatile double val =
-        (lane_id < data_size) ? static_cast<double>(scores[lane_id]) : static_cast<double>(0);
+    volatile double val = (lane_id < data_size && !is_masked(k, lane_id))
+                              ? static_cast<double>(scores[lane_id])
```
Comment (yaox12) [inline transformer_engine/common/fused_router/utils.h:187]: Ditto.

File: transformer_engine/common/fused_router/utils.h [hunk 4]
```diff
@@ -145,19 +170,29 @@ __device__ inline void apply_softmax_on_float(DataType *scores, int data_size, i
...
+                              : static_cast<double>(-1e30);
     volatile int index = (lane_id < data_size) ? lane_id : 0;
     // Some value is hanlded in local thread
     // Thread 0 is responsible for the: 0-th, 32-th, 64-th, 96-th ...
     // Reduce the value in local thread
     for (int i = lane_id + kThreadsPerWarp; i < data_size; i += kThreadsPerWarp) {
```
Comment (yaox12) [inline transformer_engine/common/fused_router/utils.h:193]: Ditto.

File: transformer_engine/common/fused_router/utils.h (hunk 1 above)
Comment (Autumn1998) [reply transformer_engine/common/fused_router/utils.h:83]: updated

**Comment by yaox12** [comment]
CI failures are irrelevant.
